Add cipher suite tests to ssl_test_old for DTLS

dwmw2 opened the pull request:
As [discussed](https://github.com/openssl/openssl/pull/1666#issuecomment-252048193), `ssl_test_old` iterates over all ciphersuites for TLS protocols but doesn't do DTLS.

I am no longer too bothered about that addition, since in #1666 I've since added a separate MTU test which will iterate over all DTLS ciphersuites anyway. But in order to get `ssl_test_old` doing it, I also had to fix `SSL_CTX_set_cipher_list()` to accept "`DTLSv1`" and "`DTLSv1.2`" as arguments. That might be worth having, so I'm submitting it separately for posterity.

levitte commented on test/recipes/80-test_ssl_old.t:
```diff
@@ -466,10 +468,15 @@ sub testssl {
+	    my $protoarg = "";
+	    $protoarg = "-ssl3" if $protocol eq "SSLv3";
+	    $protoarg = "-dtls1" if $protocol eq "DTLSv1";
+	    $protoarg = "-dtls12" if $protocol eq "DTLSv1.2";
```
I would very much prefer if you made `protoarg` a list.  The reason is that the default value will give a 0 char argument string, and there's a risk that this becomes an actual empty argument in the call below, if it gets quoted.  It's better if the default is like this, as that will just disappear:

``` perl
	    my @protoarg = ();
```

levitte replied:
Replace those four lines with this:

``` perl
	    my @protoarg = ();
 	    @protoarg = ("-ssl3") if $protocol eq "SSLv3";
 	    @protoarg = ("-dtls1") if $protocol eq "DTLSv1";
 	    @protoarg = ("-dtls12") if $protocol eq "DTLSv1.2";
```

levitte commented on test/recipes/80-test_ssl_old.t:
```diff
@@ -466,10 +468,15 @@ sub testssl {
 	    foreach my $cipher (@{$ciphersuites{$protocol}}) {
 		ok(run(test([@ssltest, @exkeys, "-cipher", $cipher,
-			     $protocol eq "SSLv3" ? ("-ssl3") : ()])),
+			     ($protoarg)])),
```
So this should be changed to:

``` perl
			     @protoarg])),
```

dwmw2 replied:
I have no idea what you just said.  Are you assuming I know perl? :)

Will attempt this.

levitte replied:
So this specific line that I'm commenting here, just change exactly like I showed you.

richsalz commented:
I think with all those parens it’s more like lisp :)

dwmw2 commented:
> As for testing in general, we would prefer if new tests are added to 80-test_ssl_new.t rather than the old.

This isn't exactly a *new* test... it is existing code in the old test, which *already* supported DTLS. It was just never *invoked* before... just because we had no way to list the ciphersuites with `openssl ciphers DTLSv1` ... which I now added. 

I *have* also added a new test which does this, in #1666.

mattcaswell commented on test/recipes/80-test_ssl_old.t:
```diff
@@ -466,10 +468,15 @@ sub testssl {
+	    @protoarg = ("-ssl3") if $protocol eq "SSLv3";
+	    @protoarg = ("-dtls1") if $protocol eq "DTLSv1";
+	    @protoarg = ("-dtls12") if $protocol eq "DTLSv1.2";
+
```
There is a non-trivial conflict in this section with current master

mattcaswell commented on apps/ciphers.c:
```diff
@@ -135,6 +143,20 @@ int ciphers_main(int argc, char **argv)
             min_version = TLS1_2_VERSION;
             max_version = TLS1_2_VERSION;
             break;
+#ifndef OPENSSL_NO_DTLS1
```
I don't think these guards are actually necessary here. We don't have them for the TLS options.

mattcaswell commented:
Just taking away the "reviewed" label for the moment...due to the above issues.

richsalz commented:
okay from me with or without the guards.  @mattcaswell still want them removed?
+1

richsalz commented:
and since it's a test, +1 for 1.1.0 too.

mattcaswell commented:
> okay from me with or without the guards. @mattcaswell still want them removed?

I'd rather avoid cluttering the code up with guards that we don't need. In any case this branch needs to have the merge conflicts resolved before we can progress.

mattcaswell commented:
Ping @dwmw2. This seems to have stalled. Are you planning on progressing this?

richsalz commented:
we don't really want things added to the old SSL test.

mattcaswell commented:
This seems to have stalled beyond the point of resurrection. Closing.
